0001-metric-pre-defined-labels.md

```diff
@@ -0,0 +1,37 @@
+# Pre-defined label support for all metric operations
+
+Let all Metric objects (Cumulative, Gauge, ...) and Raw statistics support pre-defined label values.
```

Comment:
At least from my understanding, the idea isn't that label *values* are known at compile time, but that label *keys* are known. Does that clarify things?

Reply:
That is opposite to the idea of the "raw statistics" ("delayed metrics"). The whole point to support this is to allow deferred/delayed label keys and aggregation. See for example https://github.com/census-instrumentation/opencensus-specs/blob/master/stats/gRPC.md we do record all these measurements and recommend some views, but application owners (who configured the Views) can define any label keys they want. One example is when an A/B test is performed in ServiceA (ServiceA calls ServiceB), you would want to add the `test_label` as a label key for all rpc stats (maybe just some of them) to monitor the effects. But you cannot change the gRPC instrumented code so then we need to allow this mechanism.

Reply:
That sounds like an issue with the gRPC instrumentation that can be solved with increased configurability.

Reply:
It is not about "not allowing" to override the default label names, is about adding extra labels that are propagated from different services with the context (tags/distributedContext). This is the whole idea. Also I mentioned the aggregation function (for example do I want to build a histogram with these buckets or that buckets, or just build a Sum).

If you try to come up with that "flexible configuration" you will see that you get very close to what we propose in this framework "raw statistics" + views.

Reply:
+1000 to being able to pre-specify defaults for some tags, rather than having all the tags need to be re-checked at runtime. To be all technical... this feels like the metrics version of performing partial application.